### PR TITLE
[WIP] Fix fixup_libgfortran on GCC7 problems

### PR DESCRIPTION
**Pull request by staticfloat**
This rearchitects the `contrib/fixup_libgfortran.sh` script to no longer hardcode assumptions about SONAMEs.  It instead copies _all_ SONAMEs of the three libraries we're interested in (`libgfortran`, `libgcc_s` and `libquadmath`) that it finds things like `arpack`, `libopenlibm`, etc... linking against, and deals with the names of the libraries in a much more dynamic and flexible way.

This should work for GCC 6 and GCC 7 transparently.  I'm going to run it through the current buildbots and compare against the testing buildbots (or as yichao has started calling them, the buildogs.  They're gnarly, angry, have many cutting edges, and follow in a Julian tradition of naming things after amusing autocorrect mistakes)

### CODE REVIEW

File: contrib/fixup-libgfortran.sh [hunk 1]
```diff
@@ -1,6 +1,5 @@
 #!/bin/sh
 # This file is a part of Julia. License is MIT: https://julialang.org/license
-
```
Comment (tkelman) [inline contrib/fixup-libgfortran.sh]: the license header script will put this newline back

File: contrib/fixup-libgfortran.sh [hunk 2]
```diff
@@ -40,65 +39,87 @@ find_shlib ()
...
-        LIBGFORTRAN_DIRS="$LIBGFORTRAN_DIRS $(dirname $(find_shlib $lib libgcc_s) 2>/dev/null)"
-        LIBGFORTRAN_DIRS="$LIBGFORTRAN_DIRS $(dirname $(find_shlib $lib libquadmath) 2>/dev/null)"
+        # Find the paths to the libraries we're interested in.  These are almost
+        # always within the same directory, but we like to be general.
+        LIBGFORTRAN_PATH=$(find_shlib $lib libgfortran)
+        LIBGCC_PATH=$(find_shlib $lib libgcc_s)
+        LIBQUADMATH_PATH=$(find_shlib $lib libquadmath)
+
+        # Take the directories, add them onto LIBGFORTRAN_DIRS to search for things later
+        LIBGFORTRAN_DIRS="$LIBGFORTRAN_DIRS $(dirname $LIBGFORTRAN_PATH 2>/dev/null)"
```
Comment (tkelman) [inline contrib/fixup-libgfortran.sh]: is it dirname that will be outputting to stderr, or find_shlib?

File: contrib/fixup-libgfortran.sh [hunk 3]
```diff
@@ -40,65 +39,87 @@ find_shlib ()
...
 
-LIBGFORTRAN_DIRS=$(echo "$LIBGFORTRAN_DIRS" | tr " " "\n" | sort | uniq | grep -v '^$' | tr "\n" " ")
+
+change_linkage()
+{
+    shlib_path="$1"
+    old_link="$2"
+    new_link="$3"
+    if [ "$UNAME" = "Darwin" ]; then
+        install_name_tool -change "$old_link" "$new_link" "$shlib_path" 2>/dev/null
```
Comment (tkelman) [inline contrib/fixup-libgfortran.sh]: what would cause errors?

**Comment by staticfloat** [comment]
I've moved the @jlbuild bot to a more permanent server that's not just my desktop, so I think it's ready to hit prime time.  We'll see if new issues pop up when this thing is running for weeks at a time:

@jlbuild !nuke !filter=linux64,linux32,osx
```julia
@show filter(x -> contains(x, "libgfortran"), Libdl.dllist())
```

**Comment by jlbuild** [comment]
Got it, building https://github.com/JuliaLang/julia/commit/aa9d6bad5f9537d59afd9b55a32bd67975647c51

I will edit this comment as the build jobs complete.

**Comment by jlbuild** [comment]
Got it, building https://github.com/JuliaLang/julia/commit/aa9d6bad5f9537d59afd9b55a32bd67975647c51

I will edit this comment as the build jobs complete.

**Comment by staticfloat** [comment]
The "Code Output" column is not going to finish until I add the blank testing builders.  We'll get there eventually.  I had to change some docker configs to get ipv6 loopback interfaces added in, so I'm going to run the tests again, this time with all the builders:

@jlbuild !nuke
```julia
@show filter(x -> contains(x, "libgfortran"), Libdl.dllist())
```

**Comment by jlbuild** [comment]
Got it, building https://github.com/JuliaLang/julia/commit/aa9d6bad5f9537d59afd9b55a32bd67975647c51

I will edit this comment as the build jobs complete.

**Comment by jlbuild** [comment]
Got it, building https://github.com/JuliaLang/julia/commit/aa9d6bad5f9537d59afd9b55a32bd67975647c51

I will edit this comment as the build jobs complete.

**Comment by staticfloat** [comment]
Alright, looks like I have the sparsevec problem persisting on `linux32`, and a segfault occurring on ARMv7l within `fastmath`.  Sounds like a fun Sunday afternoon.

**Comment by staticfloat** [comment]
I can confirm that this patch is working on `linux64`, `aarch64`, `ppc64le`, `osx64`; it correctly grabs `libgfortran` and friends.

**Comment by tkelman** [comment]
see review above, would be better if this script weren't suppressing errors so many places

**Comment by staticfloat** [comment]
For reference, the sparsevec problem is actually a `ccall` problem, documented here: https://github.com/JuliaLang/julia/issues/21742

File: contrib/fixup-libgfortran.sh [hunk 4]
```diff
@@ -40,65 +40,87 @@ find_shlib ()
...
+    else # $UNAME is "Linux", we only have two options, see above
+        patchelf --set-rpath \$ORIGIN "$shlib_path"
+    fi
+}
+
+LIBGFORTRAN_DIRS=$(uniquify "$LIBGFORTRAN_DIRS")
 echo "Found traces of libgfortran/libgcc in $LIBGFORTRAN_DIRS"
 
+# For every library that remotely touches libgfortran stuff (so the libraries
+# we have copied in ourselves as well as arpack, openlibm, etc...) we must
```
Comment (tkelman) [inline contrib/fixup-libgfortran.sh]: openlibm shouldn't have any fortran in it these days, that was all moved to openspecfun

**Comment by staticfloat** [comment]
@jlbuild 
```julia
@show filter(x -> contains(x, "libgfortran"), Libdl.dllist())
```

**Comment by jlbuild** [comment]
Got it, building https://github.com/JuliaLang/julia/commit/cd468eecd45230ca3aa7d22cdae244649e260ec7

I will edit this comment as the build jobs complete.

**Comment by tkelman** [comment]
the diff here is looking pretty good to me

**Comment by staticfloat** [comment]
Good, I'm running another series of tests because I downloaded an OSX binary built with GCC 7 and it wasn't quite right; it had missed `libcholmod.dylib`.  I'm going to ensure it's working on linux and osx before pulling the lever.

**Comment by staticfloat** [comment]
Alright I think we're good to go now.

**Comment by tkelman** [comment]
Did you remove the gfortran-6 symlink on the buildbots? let's make sure nightlies build okay with gcc 7 before merging the backports

**Comment by staticfloat** [comment]
I foolishly tried to build 0.5.2 bottles, which caused `brew` to remove the `gcc@6` formula, because the `julia` formula relies on `gcc` through its `:gfortran` dependency.  I think things should be back to good now.

> let's make sure nightlies build okay with gcc 7 before merging the backports

I've unfortunately already found an amendment we need to make to this before backporting.  I recently upgraded the testing infrastructure to have "tabula rasa" builders that don't have things like `gfortran` installed on them; these are catching things like incorrect `RPATH` entries setup by the `fixup-libgfortran` script, which is nice.  I'll submit a PR once I have a full run through of linux and OSX again.

**Comment by staticfloat** [comment]
ref https://github.com/JuliaLang/julia/pull/21813

**Pull request merged**